File: PythonClient/dataapi/data_api_client.py
import http.client
import ssl
import json

class DataApiClient(object):
    server_address = ""
    server_port = 443
    accesstoken = None

    def __init__(self, server_address = "", server_port = 443):
        self.server_address = server_address
        self.server_port = server_port


    def isavailable(self):
        url = self.build_url("api/servicestatus/ping")
        response, content = self._run_get_request(url)
        return response.status == 200


    def register(self, username, firstname, lastname, email, password):
        url = self.build_url( "api/account/register")
        body = {"Username" : username, "FirstName" : firstname, "LastName" : lastname, "Email" : email, "Password" : password}
        response, content = self._run_post_request(url, body = body)
        return response.status == 200


    def login(self, username, password):
        url = self.build_url( "api/account/login")
        body = {"Username" : username, "Password" : password}
        response, content = self._run_post_request(url, body = body)
        authentication_result = json.loads(content)
        if authentication_result["IsAuthenticated"]:
            self.accesstoken = authentication_result["AccessToken"]
        return authentication_result["IsAuthenticated"]


    def logout(self):
        self.accesstoken = None
        return True
    

    def changepassword(self, username, password):
        url = self.build_url( "api/account/changepassword")
        body = {"Username" : username, "Password" : password}
        response, content = self._run_post_request(url, body)
        return response.status == 200


    def deleteuser(self, username):
        url = self.build_url( "api/account/deleteuser", {"username" : username})
        response, content = self._run_delete_request(url)
        return response.status == 200


    def listcollections(self, include_hidden = False):
        url = self.build_url( "api/dataio/listcollections", {"includeHidden":include_hidden})
        response, content = self._run_get_request(url)
        return json.loads(content)


    def listcollectionnames(self, include_hidden = False):
        url = self.build_url( "api/dataio/listcollectionnames", {"includeHidden":include_hidden})
        response, content = self._run_get_request(url)
        return json.loads(content)


    def getcollectioninformation(self, collectionname):
        url = self.build_url( "api/dataio/getcollectioninformation", {"collectionName":collectionname})
        response, content = self._run_get_request(url)
        return json.loads(content)


    def getcollectionpermissions(self, collectionname):
        url = self.build_url( "api/dataio/getcollectionpermissions", {"collectionName":collectionname})
        response, content = self._run_get_request(url)
        return json.loads(content)


    def exists(self, datatype, id):
        url = self.build_url( "api/dataio/exists", {"dataType" : datatype, "id" : id})
        response, content = self._run_get_request(url)
        return response.status == 200


    def get(self, datatype, id):
        url = self.build_url( "api/dataio/get", {"dataType" : datatype, "id" : id})
        response, content = self._run_get_request(url)
        if(response.status != 200):
          raise Exception("Could not find data of type " + datatype + " and ID '" + id + "': " + response.reason + " (" + str(content) + ")")
        return json.loads(content)


    def getmany(self, datatype, query, limit = -1):
        url = self.build_url( "api/dataio/getmany", {"dataType" : datatype, "whereArguments" : query, "limit" : limit})
        response, content = self._run_get_request(url)
        if(response.status != 200):
          raise Exception("Could not get data of type " + datatype + ": " + response.reason + " (" + str(content) + ")")
        return self._read_json_lines(content)


    def search(self, query, result_format = "Json"):
        url = self.build_url( "api/dataio/search")
        body = {"Query" : query, "Format" : result_format}
        response, content = self._run_post_request(url, body)
        if(response.status != 200):
          raise Exception("Search with query '" + query + "' resulted in an error " + response.status + ": " + response.reason + " (" + str(content) + ")")
        return self._read_json_lines(content)


    def insert(self, datatype, obj, id = None):
        url = self.build_url( "api/dataio/submit")
        body = {"DataType" : datatype, "Id" : id, "Overwrite" : False, "Data" : obj}
        response, content = self._run_post_request(url, body)
        if(response.status != 200):
          raise Exception("Could not store data of type " + datatype + ": " + response.reason + " (" + str(content) + ")")
        return content


    def replace(self, datatype, obj, id = None):
        url = self.build_url( "api/dataio/submit")
        body = {"DataType" : datatype, "Id" : id, "Overwrite" : True, "Data" : obj}
        response, content = self._run_post_request(url, body)
        if(response.status != 200):
          raise Exception("Could not store data of type " + datatype + ": " + response.reason + " (" + str(content) + ")")
        return content


    def delete(self, datatype, id):
        url = self.build_url( "api/dataio/delete", {"dataType" : datatype, "id" : id})
        response, content = self._run_delete_request(url)
        return response.status == 200


    def subscribe(self, datatype, modificationTypes, filter):
        url = self.build_url( "api/subscription/subscribe")
        body = {"DataType" : datatype, "ModificationTypes" : modificationTypes, "Filter" : filter}
        response, content = self._run_post_request(url, body)
        return content


    def unsubscribe(self, subscriptionid):
        url = self.build_url( "api/subscription/unsubscribe", {"id":subscriptionid})
        response, content = self._run_get_request(url)
        return response.status == 200


    def unsubscribeall(self, datatype):
        url = self.build_url( "api/subscription/unsubscribeall", {"dataType":datatype})
        response, content = self._run_get_request(url)
        return response.status == 200


    def getsubscriptions(self, datatype = None):
        url = self.build_url( "api/subscription/getsubscriptions", {"dataType":datatype})
        response, content = self._run_get_request(url)
        return self._read_json_lines(content)


    def getsubscribedobjects(self, datatype = None):
        url = self.build_url( "api/subscription/getsubscribedobjects", {"dataType":datatype})
        response, content = self._run_get_request(url)
        return self._read_json_lines(content)


    def deletenotification(self, notificationid):
        url = self.build_url( "api/subscription/deletenotification", {"notificationId":notificationid})
        response, content = self._run_delete_request(url)
        return response.status == 200


    def reportto(self, recipient, datatype, id):
        url = self.build_url( "api/subscription/reportto", {"recipient":recipient, "dataType":datatype, "id":id})
        response, content = self._run_get_request(url)
        return response.status == 200


    def validatevalidator(self, datatype, validator):
        url = self.build_url( "api/validator/validate")
        body = {"ValidatorDefinition":{"DataType" : datatype, "ValidatorType" : "TextRules", "Ruleset" : validator}}
        response, content = self._run_post_request(url, body)
        return json.loads(content)


    def submitvalidator(self, datatype, validator):
        url = self.build_url( "api/validator/submit")
        body = {"ValidatorDefinition":{"DataType" : datatype, "ValidatorType" : "TextRules", "Ruleset" : validator}}
        response, content = self._run_post_request(url, body)
        return response.status == 200


    def getvalidator(self, validatorid):
        url = self.build_url( "api/validator/get", {"validatorId":validatorid})
        response, content = self._run_get_request(url)
        return json.loads(content)


    def listvalidators(self, collectionname = None):
        url = self.build_url( "api/validator/getall", {"collectionName":collectionname})
        response, content = self._run_get_request(url)
        return json.loads(content)


    def approvevalidator(self, validatorid):
        url = self.build_url( "api/validator/approve", {"validatorId":validatorid})
        response, content = self._run_get_request(url)
        return response.status == 200


    def deletevalidator(self, validatorid):
        url = self.build_url( "api/validator/delete", {"validatorId":validatorid})
        response, content = self._run_delete_request(url)
        return response.status == 200


    def createview(self, query, viewid = None):
        url = self.build_url( "api/view/create")
        body = {"Query":query, "ViewId":viewid}
        response, content = self._run_post_request(url, body)
        return json.loads(content)


    def getview(self, viewid, result_format = "Json", placeholders = {}):
        query_params = {"viewId":viewid, "resultFormat":result_format}
        query_params.update(placeholders)
        url = self.build_url( "api/view/get", query_params)
        response, content = self._run_get_request(url)
        return self._read_json_lines(content)


    def deleteview(self, viewid):
        url = self.build_url( "api/view/delete", {"viewId":viewid})
        response, content = self._run_delete_request(url)
        return response.status == 200


#-----------------------------------------------
# Helper methods

        
    def build_url(self, resource_path, query = {}):
        full_query = "/" + resource_path
        first_param = True
        for key in query:
            if first_param:
                full_query += "?"
                first_param = False
            else:
                full_query += "&"
            full_query += key + "=" + str(query[key])
        return full_query
        

    def _run_get_request(self, resource_path):
        connection = http.client.HTTPSConnection(self.server_address, self.server_port)
        if(self.accesstoken == None):
            connection.request("GET", resource_path);
        else:
            headers = { "Authorization" : "Bearer " + self.accesstoken }
            connection.request("GET", resource_path, headers=headers)
        response = connection.getresponse()
        content = response.read().decode('utf-8')
        connection.close()
        return response, content


    def _run_delete_request(self, resource_path):
        connection = http.client.HTTPSConnection(self.server_address, self.server_port)
        if(self.accesstoken == None):
            connection.request("DELETE", resource_path);
        else:
            headers = { "Authorization" : "Bearer " + self.accesstoken }
            connection.request("DELETE", resource_path, headers=headers)
        response = connection.getresponse()
        content = response.read().decode('utf-8')
        connection.close()
        return response, content


    def _run_post_request(self, resource_path, body):
        connection = http.client.HTTPSConnection(self.server_address, self.server_port)
        headers = { "Content-Type": "application/json" }
        if(self.accesstoken == None):
            connection.request("POST", resource_path, body=json.dumps(body), headers=headers);
        else:
            headers.update({ "Authorization" : "Bearer " + self.accesstoken })
            connection.request("POST", resource_path, body=json.dumps(body), headers=headers)
        response = connection.getresponse()
        content = response.read().decode('utf-8')
        connection.close()
        return response, content


    def _read_json_lines(self, content):
        lines = content.splitlines()
        search_result = []
        for line in lines:
          parsed_item = json.loads(line)
          search_result.append(parsed_item)
        
        return search_result

File: RClient/R/data_api_client.r
#' @export
dataapi_build_configuration <- function(server_address = "", server_port = 443){
  return( list(server_address = server_address, server_port = server_port) );
}

#' @export
dataapi_isavailable <- function(configuration)
{
	url <- build_dataapi_url(configuration,"api/servicestatus/ping");
	r <- run_get_request(url);
	return( r$status_code == 200 );
}

#' @export
dataapi_register <- function(configuration, username, firstname, lastname, email, password)
{
	url <- build_dataapi_url(configuration, "api/account/register");
	body <- list(Username = username, FirstName = firstname, LastName = lastname, Email = email, Password = password);
	r <- run_post_request(url, body = body);

	return( r$status_code == 200 );
}

#' @export
dataapi_login <- function(configuration, username, password)
{
	url <- build_dataapi_url(configuration, "api/account/login");
	body <- list(Username = username, Password = password);

	r <- run_post_request(url, body = body);
	return( httr::content(r, "parsed") );
}

#' @export
dataapi_logout <- function(configuration, accesstoken)
{
	return( TRUE );
}

#' @export
dataapi_changepassword <- function(configuration, accesstoken, username, password)
{
	url <- build_dataapi_url(configuration, "api/account/changepassword");
	body <- list(Username = username, Password = password);

	r <- run_post_request(url, body, accesstoken);
	return( r$status_code == 200 );
}

#' @export
dataapi_deleteuser <- function(configuration, accesstoken, username)
{
	url <- build_dataapi_url(configuration, "api/account/deleteuser", list(username = username));
	r <- run_delete_request(url, accesstoken);
	return( r$status_code == 200 );
}

#' @export
dataapi_listcollections <- function(configuration, accesstoken, include_hidden = FALSE)
{
	url <- build_dataapi_url(configuration, "api/dataio/listcollections", list(includeHidden=include_hidden));
	r <- run_get_request(url, accesstoken);
	return( httr::content(r, "parsed") );
}

#' @export
dataapi_listcollectionnames <- function(configuration, accesstoken, include_hidden = FALSE)
{
	url <- build_dataapi_url(configuration, "api/dataio/listcollectionnames", list(includeHidden=include_hidden));
	r <- run_get_request(url, accesstoken);
	return( httr::content(r, "parsed") );
}

#' @export
dataapi_getcollectioninformation <- function(configuration, accesstoken, collectionname)
{
	url <- build_dataapi_url(configuration, "api/dataio/getcollectioninformation", list(collectionName=collectionname));
	r <- run_get_request(url, accesstoken);
	return( httr::content(r, "parsed") );
}

#' @export
dataapi_getcollectionpermissions <- function(configuration, accesstoken, collectionname)
{
	url <- build_dataapi_url(configuration, "api/dataio/getcollectionpermissions", list(collectionName=collectionname));
	r <- run_get_request(url, accesstoken);
	return( httr::content(r, "parsed") );
}

#' @export
dataapi_exists <- function(configuration, accesstoken, datatype, id)
{
	url <- build_dataapi_url(configuration, "api/dataio/exists", list(dataType = datatype, id = id));
	r <- run_get_request(url, accesstoken);
	return( r$status_code == 200 );
}

#' @export
dataapi_get <- function(configuration, accesstoken, datatype, id)
{
	url <- build_dataapi_url(configuration, "api/dataio/get", list(dataType = datatype, id = id));
	r <- run_get_request(url, accesstoken);
	if(r$status_code != 200)
	{
	  stop(paste("Could not find data of type ", datatype, " and ID '", id, "'", sep=""));
	}
	return( httr::content(r, "parsed") );
}

#' @export
dataapi_getmany <- function(configuration, accesstoken, datatype, query, limit = NULL)
{
	url <- build_dataapi_url(configuration, "api/dataio/getmany", list(dataType = datatype, whereArguments = query, limit = limit));
	r <- run_get_request(url, accesstoken);
	if(r$status_code != 200)
	{
	  stop(paste("Could not get data of type ", datatype, sep=""));
	}
	search_result <- read_lines(httr::content(r, "text"))
	return( search_result );
}

#' @export
dataapi_search <- function(configuration, accesstoken, query, result_format = "Json")
{
	url <- build_dataapi_url(configuration, "api/dataio/search");
	body <- list(Query = query, Format = result_format);
	r <- run_post_request(url, body, accesstoken);
	if(r$status_code != 200)
	{
	  stop(paste("Search with query '", query, "' resulted in an error ", r$status_code, ": ", httr::content(r), sep=""));
	}
	search_result <- read_lines(httr::content(r, "text"));
	return( search_result );
}

#' @export
dataapi_insert <- function(configuration, accesstoken, datatype, obj, id = NULL)
{
	url <- build_dataapi_url(configuration, "api/dataio/submit");
	body <- list(DataType = datatype, Id = id, Overwrite = FALSE, Data = obj);
	r <- run_post_request(url, body, accesstoken);
	if(r$status_code != 200)
	{
	  stop(paste("Could not store data of type ", datatype, ": ", httr::content(r), sep=""));
	}
	return( httr::content(r, "text", encoding = "UTF-8") );
}

#' @export
dataapi_replace <- function(configuration, accesstoken, datatype, obj, id = NULL)
{
	url <- build_dataapi_url(configuration, "api/dataio/submit");
	body <- list(DataType = datatype, Id = id, Overwrite = TRUE, Data = obj);
	r <- run_post_request(url, body, accesstoken);
	if(r$status_code != 200)
	{
	  stop(paste("Could not store data of type ", datatype, ": ", httr::content(r), sep=""));
	}
	return( httr::content(r, "text", encoding = "UTF-8") );
}

#' @export
dataapi_delete <- function(configuration, accesstoken, datatype, id)
{
	url <- build_dataapi_url(configuration, "api/dataio/delete", list(dataType = datatype, id = id));
	r <- run_delete_request(url, accesstoken);
	return( r$status_code == 200 );
}

#' @export
dataapi_subscribe <- function(configuration, accesstoken, datatype, modificationTypes, filter = "")
{
	url <- build_dataapi_url(configuration, "api/subscription/subscribe");
	body <- list(DataType = datatype, ModificationTypes = modificationTypes, Filter = filter);
	r <- run_post_request(url, body, accesstoken);
	return( httr::content(r, "text", encoding = "UTF-8") );
}

#' @export
dataapi_unsubscribe <- function(configuration, accesstoken, subscriptionid)
{
	url <- build_dataapi_url(configuration, "api/subscription/unsubscribe", list(id=subscriptionid));
	r <- run_get_request(url, accesstoken);
	return( r$status_code == 200 );
}

#' @export
dataapi_unsubscribeall <- function(configuration, accesstoken, datatype)
{
	url <- build_dataapi_url(configuration, "api/subscription/unsubscribeall", list(dataType=datatype));
	r <- run_get_request(url, accesstoken);
	return( r$status_code == 200 );
}

#' @export
dataapi_getsubscriptions <- function(configuration, accesstoken, datatype = NULL)
{
	url <- build_dataapi_url(configuration, "api/subscription/getsubscriptions", list(dataType=datatype));
	r <- run_get_request(url, accesstoken);
	subscriptions <- read_lines(httr::content(r, "text"))
	return( subscriptions );
}

#' @export
dataapi_getsubscribedobjects <- function(configuration, accesstoken, datatype = NULL)
{
	url <- build_dataapi_url(configuration, "api/subscription/getsubscribedobjects", list(dataType=datatype));
	r <- run_get_request(url, accesstoken);
	notifications <- read_lines(httr::content(r, "text"))
	return( notifications  );
}

#' @export
dataapi_deletenotification <- function(configuration, accesstoken, notificationid)
{
	url <- build_dataapi_url(configuration, "api/subscription/deletenotification", list(notificationId=notificationid));
	r <- run_delete_request(url, accesstoken);
	return( r$status_code == 200 );
}

#' @export
dataapi_reportto <- function(configuration, accesstoken, recipient, datatype, id)
{
	url <- build_dataapi_url(configuration, "api/subscription/reportto", list(recipient=recipient, dataType=datatype, id=id));
	r <- run_get_request(url, accesstoken);
	return( r$status_code == 200 );
}

#' @export
dataapi_validatevalidator <- function(configuration, accesstoken, datatype, validator)
{
	url <- build_dataapi_url(configuration, "api/validator/validate");
	body <- list(ValidatorDefinition=list(DataType = datatype, ValidatorType = "TextRules", Ruleset = validator));
	r <- run_post_request(url, body, accesstoken);
	return( httr::content(r, "parsed") );
}

#' @export
dataapi_submitvalidator <- function(configuration, accesstoken, datatype, validator)
{
	url <- build_dataapi_url(configuration, "api/validator/submit");
	body <- list(ValidatorDefinition=list(DataType = datatype, ValidatorType = "TextRules", Ruleset = validator));
	r <- run_post_request(url, body, accesstoken);
	return( r$status_code == 200 );
}

#' @export
dataapi_getvalidator <- function(configuration, accesstoken, validatorid)
{
	url <- build_dataapi_url(configuration, "api/validator/get", list(validatorId=validatorid));
	r <- run_get_request(url, accesstoken);
	return( httr::content(r, "parsed") );
}

#' @export
dataapi_listvalidators <- function(configuration, accesstoken, collectionname = NULL)
{
	url <- build_dataapi_url(configuration, "api/validator/getall", list(collectionName=collectionname));
	r <- run_get_request(url, accesstoken);
	return( httr::content(r, "parsed") );
}

#' @export
dataapi_approvevalidator <- function(configuration, accesstoken, validatorid)
{
	url <- build_dataapi_url(configuration, "api/validator/approve", list(validatorId=validatorid));
	r <- run_get_request(url, accesstoken);
	return( r$status_code == 200 );
}

#' @export
dataapi_deletevalidator <- function(configuration, accesstoken, validatorid)
{
	url <- build_dataapi_url(configuration, "api/validator/delete", list(validatorId=validatorid));
	r <- run_delete_request(url, accesstoken);
	return( r$status_code == 200 );
}

#' @export
dataapi_createview <- function(configuration, accesstoken, query, viewid = NULL)
{
	url <- build_dataapi_url(configuration, "api/view/create");
	body <- list(Query=query, ViewId=viewid);
	r <- run_post_request(url, body, accesstoken);
	return( httr::content(r, "parsed") );
}

#' @export
dataapi_getview <- function(configuration, accesstoken, viewid, result_format = "Json")
{
	url <- build_dataapi_url(configuration, "api/view/get", list(viewId=viewid, resultFormat=result_format));
	r <- run_get_request(url, accesstoken);
	search_result <- read_lines(httr::content(r, "text"))
	return( search_result  );
}

#' @export
dataapi_deleteview <- function(configuration, accesstoken, viewid)
{
	url <- build_dataapi_url(configuration, "api/view/delete", list(viewId=viewid));
	r <- run_delete_request(url, accesstoken);
	return( r$status_code == 200 );
}

#' @export
dataapi_addtodataproject <- function(configuration, accesstoken, dataprojectid, datatype, dataid, initials, filename = NULL)
{
  upload_id <- uuid::UUIDgenerate();
  current_utc_time <- as.POSIXlt(Sys.time(), tz = "UTC")
  data_project_upload_info <- list(Id = upload_id,
                                   UploaderInitials = initials,
                                   UploadTimestamp = format(current_utc_time, "%Y-%m-%dT%H:%M:%SZ"),
                                   DataProjectId = dataprojectid,
                                   Filename = filename,
                                   RawData = list(
                                     DataType = datatype,
                                     Id = dataid
                                   ),
                                   DerivedData = c()
                               )
  dataapi::dataapi_insert(configuration, accesstoken, 'DataProjectUploadInfo', data_project_upload_info, upload_id)
  return( upload_id );
}

build_dataapi_url <- function(configuration, path, query = list())
{
	full_path <- paste("https://", configuration$server_address, ":", configuration$server_port, "/", path, sep="");
	if(length(query) > 0)
	{
		query_str <- "";
		for(parameterName in ls(query))
		{
			parameterValue <- query[[parameterName]];
			if(!is.null(parameterValue)) {
			  parameterValue <- utils::URLencode(toString(parameterValue));
			  if(query_str != ""){
			    query_str <- paste(query_str, "&", sep="");
			  }
			  query_str <- paste(query_str, paste(parameterName, parameterValue, sep="="), sep = "");
			}
		}
		full_path <- paste(full_path, "?", query_str, sep="");
	}
	return( full_path );
}

run_get_request <- function(url, accesstoken = NULL)
{
  if(is.null(accesstoken))
  {
    return( httr::GET(url));
  }
  else
  {
    return( httr::GET(url, httr::add_headers(Authorization = paste("Bearer", accesstoken, sep = " "))) );
  }
}

run_delete_request <- function(url, accesstoken = NULL)
{
  if(is.null(accesstoken))
  {
    return( httr:: DELETE(url));
  }
  else
  {
    return( httr::DELETE(url, httr::add_headers(Authorization = paste("Bearer", accesstoken, sep = " "))) );
  }
}

run_post_request <- function(url, body, accesstoken = NULL)
{
  if(is.null(accesstoken))
  {
    return( httr::POST(url, body = rjson::toJSON(body), httr::add_headers(`Content-Type` = "application/json")));
  }
  else
  {
    return( httr::POST(url, body = rjson::toJSON(body), httr::add_headers(`Content-Type` = "application/json", Authorization = paste("Bearer", accesstoken, sep = " "))) );
  }
}

read_lines <- function(content)
{
  lines <- strsplit(content, "\n")[[1]]
  search_result <- list();
  item_index <- 0;
  for(line in lines)
  {
    item_index <- item_index + 1;
    parsed_item <- rjson::fromJSON(line);
    search_result[[item_index]] <- parsed_item;
  }
  return(search_result)
}
